File: ChemLabBackend/src/ChemLab.Domain/Entities/Request.cs
using System;
using System.Collections.Generic;

namespace ChemLab.Domain.Entities;

public class Request
{
    public Guid Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public RequestType Type { get; set; }
    public RequestStatus Status { get; set; }
    public RequestPriority Priority { get; set; }
    public DateTime RequestedAt { get; set; }
    public DateTime? ExpectedDate { get; set; }
    public DateTime? ResolvedAt { get; set; }
    public string? Comments { get; set; }
    public string? RejectionReason { get; set; }

    // Relaciones
    public Guid RequestedById { get; set; }
    public virtual User RequestedBy { get; set; } = null!;

    public Guid? ApprovedById { get; set; }
    public virtual User? ApprovedBy { get; set; }

    public virtual ICollection<RequestItem> Items { get; set; } = new List<RequestItem>();
}

public enum RequestType
{
    Reagent = 0,
    Equipment = 1,
    Maintenance = 2,
    Calibration = 3,
    Other = 4
}

public enum RequestStatus
{
    Pending = 0,
    Approved = 1,
    Rejected = 2,
    InProgress = 3,
    Completed = 4,
    Cancelled = 5
}

public enum RequestPriority
{
    Low = 0,
    Medium = 1,
    High = 2,
    Urgent = 3
}

public class RequestItem
{
    public Guid Id { get; set; }
    public Guid RequestId { get; set; }
    public virtual Request Request { get; set; } = null!;

    public string ItemName { get; set; } = string.Empty;
    public string? CatalogNumber { get; set; }
    public string? CasNumber { get; set; }
    public decimal? Quantity { get; set; }
    public string? Unit { get; set; }
    public string? Specifications { get; set; }
    public string? Supplier { get; set; }
    public decimal? EstimatedPrice { get; set; }
}

File: frontend/src/types/request.ts
export type RequestType = 'Reagent' | 'Equipment' | 'Maintenance' | 'Calibration' | 'Other';
export type RequestStatus = 'Pending' | 'Approved' | 'Rejected' | 'InProgress' | 'Completed' | 'Cancelled';
export type RequestPriority = 0 | 1 | 2 | 3;

export interface RequestItem {
  id?: string;
  itemName: string;
  catalogNumber?: string;
  casNumber?: string;
  quantity?: number;
  unit?: string;
  specifications?: string;
  supplier?: string;
  estimatedPrice?: number;
}

export interface Request {
  id: string;
  title: string;
  description: string;
  type: RequestType;
  status: RequestStatus;
  priority: RequestPriority;
  requestedAt: string;
  expectedDate?: string;
  resolvedAt?: string;
  comments?: string;
  rejectionReason?: string;
  requestedBy: {
    id: string;
    name: string;
    email: string;
  };
  approvedBy?: {
    id: string;
    name: string;
  };
  items: RequestItem[];
  itemsCount?: number;
  totalItems?: number;
}

export interface CreateRequestDto {
  title: string;
  description: string;
  type: RequestType;
  priority: RequestPriority;
  expectedDate?: string;
  items: RequestItem[];
}

export interface ApproveRequestDto {
  comments?: string;
}

export interface RejectRequestDto {
  rejectionReason: string;
}

export const requestTypeColors: Record<RequestType, string> = {
  Reagent: '#4caf50',      // Verde
  Equipment: '#2196f3',     // Azul
  Maintenance: '#ff9800',   // Naranja
  Calibration: '#9c27b0',   // Morado
  Other: '#9e9e9e'          // Gris
};

export const requestStatusColors: Record<RequestStatus, string> = {
  Pending: '#ff9800',       // Naranja
  Approved: '#4caf50',      // Verde
  Rejected: '#f44336',      // Rojo
  InProgress: '#2196f3',    // Azul
  Completed: '#2e7d32',     // Verde oscuro
  Cancelled: '#9e9e9e'      // Gris
};

export const requestPriorityColors: Record<RequestPriority, string> = {
  [0]: '#8bc34a',           // Verde claro - Baja
  [1]: '#ff9800',           // Naranja - Media
  [2]: '#f44336',           // Rojo - Alta
  [3]: '#9c27b0'            // Morado - Urgente
};

export const requestTypeIcons: Record<RequestType, string> = {
  Reagent: '🧪',
  Equipment: '🔧',
  Maintenance: '🛠️',
  Calibration: '📏',
  Other: '📦'
};

export const requestStatusIcons: Record<RequestStatus, string> = {
  Pending: '⏳',
  Approved: '✅',
  Rejected: '❌',
  InProgress: '⚙️',
  Completed: '🎉',
  Cancelled: '🚫'
};
